feat: модификация основных файлов проекта под первую итерацию

Welpodron opened the pull request.

Welpodron committed:
feat: модификация основных файлов проекта под первую итерацию

olgalatkina commented on src/components/arrow-button/ArrowButton.stories.tsx:
```diff
@@ -13,7 +13,8 @@ export const ArrowButtonStory: Story = {
 		return (
 			<>
-				<ArrowButton />
+				<ArrowButton isActive={true} onClick={() => alert('клик на кнопку')} />
```
_надо исправить:_
Нет необходимости писать компонент дважды.

olgalatkina commented on src/components/arrow-button/ArrowButton.tsx:
```diff
@@ -2,18 +2,28 @@ import arrow from 'src/images/arrow.svg';
 			tabIndex={0}
-			className={styles.container}>
-			<img src={arrow} alt='иконка стрелочки' className={styles.arrow} />
+			className={clsx(styles.container, isActive && styles.container_open)}>
```
Отлично, что воспользовались библиотекой `clsx`!

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,48 @@
 			}>
-			<ArticleParamsForm />
+			<ArticleParamsForm
+				defaultAppState={defaultAppState}
```
_надо исправить:_
Не стоит передавать в пропсах то, что можно будет в `ArticleParamsForm` импортировать из `constants/articleProps.ts`

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,48 @@
+			<ArticleParamsForm
+				defaultAppState={defaultAppState}
+				onSubmit={setAppState}
+				onReset={setAppState}
```
_надо исправить:_
Одна и та же функция в двух пропсах - избыточно и может запутать. Передайте один проп `setAppState={setAppState}`, этого будет достаточно.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,109 @@
+export type ArticleParamsFormCallbackAction = (value: ArticleStateType) => void;
+
+export type ArticleParamsFormProps = {
+	defaultAppState: ArticleStateType;
```
_надо исправить:_
Не стоит передавать в пропсах то, что можно импортировать из `constants/articleProps.ts`

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,109 @@
+export type ArticleParamsFormProps = {
+	defaultAppState: ArticleStateType;
+	onSubmit: ArticleParamsFormCallbackAction;
+	onReset: ArticleParamsFormCallbackAction;
```
_надо исправить:_
Одна и та же функция в двух пропсах - избыточно и может запутать. В этом компоненте уже не очевидно, что в 2-х пропсах приходит одна и та же функция.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,109 @@
+			<aside
+				className={clsx(styles.container, isOpened && styles.container_open)}>
+				<form
+					onSubmit={handleSubmit}
```
Отлично, что событие `submit` навешано на форму.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,109 @@
+
+	const [formState, setFormState] = useState<ArticleStateType>(defaultAppState);
+
+	const handleChange = (fieldName: string) => {
```
Интересное решение с общим обработчиком `onChange`.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,48 @@
 const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
+export const defaultAppState: ArticleStateType = {
```
_надо исправить:_
Эту переменную нужно вынести в файл с константами и оттуда имопртировать в те файлы, где она нужна.

olgalatkina reviewed (changes requested):
Отличная продуманная работа:
- Состояния страницы и формы чётко разделены
- Использована библиотека `clsx`
- Проект хорошо типизирован
- События `submit` и `reset` навешаны на форму

Но есть несколько моментов, которые _надо поправить:_
- [ ] По ТЗ сайдбар должен закрываться по клику вне сайдбара: https://yadi.sk/d/xodZCby4QsdoFw
- [ ] В `ArticleParamsForm` достаточно передавать только сеттер нового состояния страницы

_можно лучше:_
- [ ]  Вынести компонент App в отдельный файл
- [ ]  Для одиночных селекторов не использовать библиотеку `clsx` (см. компонент `Article`)
- [ ] Использовать кастомный хук для закрытия сайдбара по клику вне сайдбара

Welpodron committed:
fix: фикс нюансов связанных с первой итерацией

Добавлена поддержка закрытия через оверлей (хук не реализован так как есть проблема связанная с
работой поставляемого по умолчанию компонента Select, в котором при закрытии попапа DOM элементы
списка пропадают из потомков ArticleParamsForm), App вынесен в отдельный компонент, уменьшено
количество пропсов ArticleParamsForm, константные значения по умолчанию берутся теперь из constants

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,110 @@
+			/>
+			<div
+				onClick={() => setIsOpened(false)}
+				className={clsx(styles.overlay, isOpened && styles.overlay_open)}></div>
```
Это самое экономное решение. Браво!

Но если хотите, можете попробовать вернуться к прежнему состоянию проекта и применить кастомный хук `useClose`, в нём так же реализовано закрытие по Escape:

```
import { useEffect } from "react";

type TUseClose = {
  isOpen: boolean;
  onClose: () => void;
  rootRef: React.RefObject<HTMLElement>;
};

// Кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. 
// Он следит за их вызовами
export function useClose({isOpen, onClose, rootRef}: TUseClose) {
  useEffect(() => {
    if (!isOpen) return; // останавливаем действие эффекта, если закрыто
   
    function handleClickOutside(event: MouseEvent) {
      const { target } = event;
      
      const isOutsideClick =
        target instanceof Node && // проверяем, что это `DOM`-элемент
        rootRef.current &&
        !rootRef.current.contains(target); // проверяем, что кликнули на элемент, который находится не внутри нашего блока
      if (isOutsideClick) {
        onClose();
      }
    }

    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    document.addEventListener("mousedown", handleClickOutside);

    //  Обязательно удаляем обработчики в `clean-up`- функции
    return () => {
      document.removeEventListener("keydown", handleEscape);
      document.removeEventListener("mousedown", handleClickOutside);
    };
    
    // Обязательно следим за `isOpen`, чтобы срабатывало только при открытии, 
    // а не при любой перерисовке компонента
  }, [isOpen, onClose, rootRef]);
}
```

Этот код помещают в отдельный файл `useClose.ts` в папке `hooks` и импортируют хук туда, где нужно.
И теперь можно использовать в любом компоненте (попапе, меню и тд), которому нужно установить эти обработчики.

Пример того, как это использовать:

```
    // Создаем `ref` для нашей формы и указываем его на внешнем блоке с формой
    const formRef = useRef<HTMLElement>(null);
    
    useClose({
        isOpen,
        onClose: () => setIsOpen(false),
        rootRef: formRef,
    });

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,110 @@
+} from 'src/constants/articleProps';
+import { RadioGroup } from '../radio-group';
+import { Separator } from '../separator';
+import { useState, FormEvent } from 'react';
```
_можно лучше:_
Иногда линтеры настраивают проверять порядок импортов.  Но даже если линтер не настроен, то стоит придерживаться какой-то логики в порядке импортов, например:
1.  всё, что импортируем из библиотеки `react`
2.  из библиотек, начинающихся со слова `react` (`react-router-dom` и т.п.)
3.  из других библиотек (`clsx`)
4.  компоненты сторонних библиотек
5.  константы из папки `constants`, вспомогательные функции из папки `utils`, стилевые файлы и т.д.
6.  свои компоненты из папки `components`

olgalatkina reviewed (approved):
Функционал реализован полностью, работа принимается!

_можно лучше:_
- [ ] Попробовать применить кастомный хук `useClose`  (см. комментарий в коде).
- [ ] Навести порядок в импортах в файле `ArticleParamsForm.tsx`
